corehq/ex-submodules/pillowtop/pillow/interface.py

```diff
@@ -100,29 +100,94 @@ def run(self):
+        # chunked processing is only supported for pillows with single processor
+        return (len(self.processors) == 1
+            and getattr(self.processors[0], 'processor_chunk_size', 0) > 0)
+
     def process_changes(self, since, forever):
         """
-        Process changes from the changes stream.
+        Handle changes stream from the changes stream and pass it to processors
         """
         context = PillowRuntimeContext(changes_seen=0)
         try:
+            changes_chunk = []
             for change in self.get_change_feed().iter_changes(since=since or None, forever=forever):
                 if change:
                     context.changes_seen += 1
-                    self.process_with_error_handling(change, context)
+                    if self._should_process_in_chunks:
+                        changes_chunk.append(change)
+                        if len(changes_chunk) == self.processors[0].processor_chunk_size:
+                            self.process_chunk_with_error_handling(changes_chunk, context)
+                            changes_chunk = []
+                    else:
+                        self.process_with_error_handling(change, context)
                 else:
                     self._update_checkpoint(None, None)
+            self.process_chunk_with_error_handling(changes_chunk, context)
         except PillowtopCheckpointReset:
+            # finish processing any ramining chunk
+            self.process_chunk_with_error_handling(changes_chunk, context)
             self.process_changes(since=self.get_last_checkpoint_sequence(), forever=forever)
 
-    def process_with_error_handling(self, change, context):
+    def process_chunk_with_error_handling(self, changes_chunk, context):
+        """
+        Passes given changes_chunk to the processor for chunked processing
+            If there is an exception in chunked processing, falls back
+            to passing changes one by one to the processor
+        """
+        if not changes_chunk:
+            return
+        failed_changes = set()
+        try:
+            # chunked processing is supported if there is only one processor
+            failed_changes = self.processors[0].process_changes_chunk(self, changes_chunk)
+        except Exception as ex:
+            notify_exception(
+                None,
+                "{pillow_name} Error in processing changes chunk {change_ids}: {ex}".format(
+                    pillow_name=self.get_name(),
+                    change_ids=[c.id for c in changes_chunk],
+                    ex=ex
+                ))
+            # fall back to processing one by one
+            for change in changes_chunk:
+                self.process_with_error_handling(change, context, chunked_fallback=True)
+        else:
+            succeeded_changes = set(changes_chunk) - set(failed_changes)
+            for change in succeeded_changes:
+                self._update_checkpoint(change, context)
```

Comment:
Is it okay to update checkpoints in this manner in chunked processing?

Reply:
My comment on this seems to been misplaced:

I haven't thought about this much but does this checkpointing process still work when doing in bulk? For couch it will just take the sequence from the change object and for kafka it will take it from the change feed. Certainly for Kafka it doesn't seem to make sense to call this for every successful change since the offset of the change feed will be the same.

There's also the question of whether we should be recording the current offset or the max offset before we hit an error. I think what we want is to make sure we don't set the checkpoint to any value beyond which we've successfully processed a change (or successfully handled the error).